Improve linspace decomposition and remove its lowering

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91625
* __->__ #91621

The code produced by the lowering and the decomposition is now the same
modulo a casting to `float32`. This casting is necessary as otherwise
the tests do not pass due to accuracy errors. We prefer accuracy over
speed here, given that this is an associative scan, and thus it's prone
to numerical errors.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

lezcano committed:
Improve linspace decomposition and remove its lowering

The code produced by the lowering and the decomposition is now the same
modulo a casting to `float64`. This casting is necessary as otherwise
the tests do not pass due to accuracy errors. We prefer accuracy over
speed here, given that this is an associative scan, and thus it's prone
to numerical errors.

[ghstack-poisoned]

lezcano committed:
Update on "Improve linspace decomposition and remove its lowering"


The code produced by the lowering and the decomposition is now the same
modulo a casting to `float64`. This casting is necessary as otherwise
the tests do not pass due to accuracy errors. We prefer accuracy over
speed here, given that this is an associative scan, and thus it's prone
to numerical errors.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
The test script
```python
import torch

def f():
    return torch.linspace(3, 10, steps=5, device="cuda")

compiled_model = torch.compile(f)
compiled_model()
```

generates in master
```python
    xnumel = 5
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = 1.75
    tmp1 = x0
    tmp2 = tmp0 * tmp1
    tmp3 = 3
    tmp4 = tmp2 + tmp3
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp4, xmask)
```

and in this PR
```python
    xnumel = 5
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = x0
    tmp1 = tmp0.to(tl.float32)
    tmp2 = 1.75
    tmp3 = tmp1 * tmp2
    tmp4 = 3
    tmp5 = tmp3 + tmp4
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp5, xmask)
```
The only difference is that the cast to float32 is explicit rather than implicit. This cast should probably be removed by a finer version of @eellison's strength reduction algorithm.

Note that the lowering had severe errors on low-precision dtypes. The current implementation fixes that.

@jancervenka have a look in case you want to have a go at removing some other lowerings while keeping the generated code the same.

lezcano committed:
Update on "Improve linspace decomposition and remove its lowering"


The code produced by the lowering and the decomposition is now the same
modulo a casting to `float64`. This casting is necessary as otherwise
the tests do not pass due to accuracy errors. We prefer accuracy over
speed here, given that this is an associative scan, and thus it's prone
to numerical errors.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

peterbell10 commented on torch/_refs/__init__.py:
```diff
@@ -4328,20 +4328,21 @@ def linspace(
+    )
+
+    cast = partial(torch.full, (), dtype=computation_dtype, **factory_kwargs)
+    out = rg * cast((end - start) / (steps - 1)) + cast(start)
```
Should this not use the algorithm from ATen:

```
where(ind < halfway, start + step * rg, end - step * (steps - rg - 1))
```

This should get the end points of the range exactly.

lezcano replied:
Sure thing. Let's kill two birds with one stone.

lezcano replied:
https://github.com/pytorch/pytorch/pull/91625

ngimel replied:
Why do we need to cast `end` and `start` to tensor? Inductor remedies this, but it still looks unnecessary?

lezcano replied:
Sure, let me cast `rg` instead.

lezcano committed:
Update on "Improve linspace decomposition and remove its lowering"


The code produced by the lowering and the decomposition is now the same
modulo a casting to `float64`. This casting is necessary as otherwise
the tests do not pass due to accuracy errors. We prefer accuracy over
speed here, given that this is an associative scan, and thus it's prone
to numerical errors.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
Yeah, sorry for not updating the PR. Now it should be fine. Also, note that the next PR in the stack solves the issue with the ends of linspace.

lezcano committed:
Update on "Improve linspace decomposition and remove its lowering"


The code produced by the lowering and the decomposition is now the same
modulo a casting to `float32`. This casting is necessary as otherwise
the tests do not pass due to accuracy errors. We prefer accuracy over
speed here, given that this is an associative scan, and thus it's prone
to numerical errors.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Improve linspace decomposition and remove its lowering"


The code produced by the lowering and the decomposition is now the same
modulo a casting to `float32`. This casting is necessary as otherwise
the tests do not pass due to accuracy errors. We prefer accuracy over
speed here, given that this is an associative scan, and thus it's prone
to numerical errors.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
@ngimel now the two PRs are ready for review.
